cfpb/sbl-frontend
# feat(Upload): Part 1 - Skeleton, API Request, Upload Status, Validation Status

closes #302 
closes #305 
closes #307 
closes #344 

## Changes
- chore(.env): utilizes the `SBL_FILING_BASE_URL` env
- chore(TypeScript): resolves TypeScript types  - FilingStatus, Submissions, Uploads
- chore(filingTypes): moves associated filingTypes to the `types/filingTypes`
- feat(Upload, Submission): Implemented `useGetSubmissionLatest` and `useUploadMutation` as well as associated Axios requests
- component(InlineStatus): Used for the Icon - Description components used in Upload
- **feat(Upload): Uploading `.csv` is functioning properly**
- feat(Upload): Fetching previous uploads upon load of an `Upload` page

## Additional Changes (03/30/2024)
- dependency(design-system-react): Updated the `design-system-react` package to make use of the Icon accessbility
- development(axiosInstance): Updated TypeScript for API requests
- development(axiosInstance): Allowed to the Request function to allow additional Axios [options](https://axios-http.com/docs/req_config)
- component(InstitutionHeading): Separated from `InstitutionCard` into its own component -- `Test Bank 456 | 2024`
- **feat(Upload - Routing): receive the financial institution's `name` from /filing page using react-router link state**
- **feat(Upload - Status): Added proper icons and message alerting the user of the Upload/Validation statuses**
- feat(Upload - Progress): Added placeholder function for `onUploadProgress` if a progress bar is to be used during upload
- **feat(Upload - Validation): Added callback that runs the Validation request after a successful upload**
- **feat(Upload - Validation): Allotted five `retries` with an exponential decay `retryDelay`**
- style(stepIndicator): placed above FormWrapper -- max-width 1170px to match the figma

## How to Test
1. Make sure `SBL_REGTECH_FILING_URL` is set properly in your `.env`
2. `yarn up -i design-system-react` to get the latest `design-system-react` dependency
3. Run `yarn update`, reset all Docker volumes and containers, and set the user to have a LEI in Keycloak
4. Navigate to `http://localhost:8899/filing` and click on 'Upload' to navigate to the lei's specific Upload page
5. Upload a `.csv` file -- can use test .csv files from https://github.com/cfpb/regtech-data-validator/tree/main/tests/data
6. Notice Upload/Validation icons and content changes; inspect the Network tab to verify API request results
7. Verify the implemented control flow listed below:

## [WIP] Control Flow Testing
- [ ] if the lei and/or filing period is not associated with the user, the user should be redirected to `/filing`
- [x] If a filing has not been posted, the entire "Select a file to upload" Section should not appear
- [x] If a previous upload was completed previously, the upload section's content and button label should change
- [x] Validation should occur after an upload
- [ ] ...Still more

## Screenshots (03/30/2024)
|Initial|Error| 
|--|--|
|<img width="815" alt="initial" src="https://github.com/cfpb/sbl-frontend/assets/13324863/94f78108-41bf-4c15-884e-d9142799d176">|<img width="821" alt="Error" src="https://github.com/cfpb/sbl-frontend/assets/13324863/079891bf-5daa-46d3-8c42-94d380215b9f">|

|Validation In Progress|Complete|
|--|--|
|<img width="806" alt="Validation in progress" src="https://github.com/cfpb/sbl-frontend/assets/13324863/19671d56-cf41-4ef5-9aac-9099b081025a">|<img width="807" alt="Complete" src="https://github.com/cfpb/sbl-frontend/assets/13324863/74fa8b2a-56e2-4d22-bed4-0deeca35f5e3">|
